kernel/src/main/java/org/kframework/unparser/KPrint.java

```diff
@@ -157,23 +157,30 @@ public void prettyPrint(Definition def, Module module, Consumer<byte[]> print, K
     }
 
     public byte[] prettyPrint(Definition def, Module module, K orig, Sort s, ColorSetting colorize, OutputModes outputMode) {
-        K result = abstractTerm(module, orig);
+        RuleGrammarGenerator gen = new RuleGrammarGenerator(def);
         switch (outputMode) {
             case KAST:
             case NONE:
             case BINARY:
             case JSON:
-            case LATEX:
+            case LATEX: {
+                Module unparsingModule = gen.getRuleGrammar(module);
+                K result = abstractTerm(unparsingModule, orig);
                 return serialize(result, outputMode);
-            case PRETTY:
-                Module prettyUnparsingModule = RuleGrammarGenerator.getCombinedGrammar(module, false).getExtensionModule();
+            }
+            case PRETTY: {
+                Module prettyUnparsingModule = RuleGrammarGenerator.getCombinedGrammar(gen.getRuleGrammar(module), false).getExtensionModule();
+                K result = abstractTerm(prettyUnparsingModule, orig);
                 return (unparseTerm(result, prettyUnparsingModule, colorize) + "\n").getBytes();
+            }
             case PROGRAM: {
-                RuleGrammarGenerator gen = new RuleGrammarGenerator(def);
                 Module programUnparsingModule = RuleGrammarGenerator.getCombinedGrammar(gen.getProgramsGrammar(module), false).getParsingModule();
+                K result = abstractTerm(programUnparsingModule, orig);
                 return (unparseTerm(result, programUnparsingModule, colorize) + "\n").getBytes();
             }
-            case KORE:
+            case KORE: {
+                Module programUnparsingModule = RuleGrammarGenerator.getCombinedGrammar(gen.getProgramsGrammar(module), false).getParsingModule();
+                K result = abstractTerm(programUnparsingModule, orig);
```

Comment:
From discussion in meeting:

This may be the wrong grammar (`gen.getProgramsGrammar(module)`)?

Likely with KORE output we want to preserve the KORE input as much as possible.

But this receives as input a KAST term.

If we want at fast path Kore -> Kore, it probably shouldn't go through this function at all.